### PR TITLE
Remove noexcepts

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

Removes noexcept specifiers and operators.
Adds a test to prevent noexcepts from being added to the code.

This is the first step to enabling better exception handling by throwing exceptions.
This will allow better compatibility with python as well as trying to have a cleaner shutdown when an error occurs during a simulation.

### Upgrade instructions

Remove all noexcept specifiers.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/DataStructures/DataBox/DataBox.hpp [hunk 1]
```diff
@@ -212,13 +212,8 @@ class DataBox<tmpl::list<Tags...>> : private detail::Item<Tags>... {
...
   DataBox() = default;
-  DataBox(DataBox&& rhs) noexcept(
-      tmpl2::flat_all_v<
-          std::is_nothrow_move_constructible_v<detail::Item<Tags>>...>) =
-      default;
-  DataBox& operator=(DataBox&& rhs) noexcept(
-      tmpl2::flat_all_v<
-          std::is_nothrow_move_assignable_v<detail::Item<Tags>>...>) {
+  DataBox(DataBox&& rhs) = default;
+  DataBox& operator=(DataBox&& rhs) {
```
Comment (nilsvu) [inline src/DataStructures/DataBox/DataBox.hpp:216]: I think this is overall a very good change. However, I don't think we should prohibit noexcept specifiers altogether. My suggestion is to follow Google's style guide:

https://google.github.io/styleguide/cppguide.html#noexcept

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/DataStructures/DataBox/DataBox.hpp (hunk 1 above)
Comment (knelli2) [reply src/DataStructures/DataBox/DataBox.hpp:216]: Is it possible to add a git hook that checks for `noexcept`s and prints a warning if you try to commit a change that has them? That way you don't prohibit it altogether, but it's much harder to accidentally add them.

Comment (kidder) [reply src/DataStructures/DataBox/DataBox.hpp:216]: > I think this is overall a very good change. However, I don't think we should prohibit noexcept specifiers altogether. My suggestion is to follow Google's style guide:
> 
> https://google.github.io/styleguide/cppguide.html#noexcept

If you are suggesting that I individually examine the 15,437 occurrences of noexcept, that will not be happening...
Note that Google doesn't allow exceptions which is the opposite of what we are moving towards.

I am okay with dropping the file check if we come up with a reasonable policy for when to use noexcept

Comment (nilsvu) [reply src/DataStructures/DataBox/DataBox.hpp:216]: Hehe of course I'm not asking you to check all noexcepts :D My suggestion:

- Drop the commit that prohibits adding noexcept
- In the docs just delete the part on noexcepts. We point to the Google style guide already.
- In the few occurrences where we implement the move constructor and move assignment, keep the noexcept (first few commits).
- Perhaps we should run quick sanity check before / after this change to make sure it doesn't have an unexpected performance hit?

Comment (kidder) [reply src/DataStructures/DataBox/DataBox.hpp:216]: @wthrowe @nilsdeppe and I discussed this and our plan is merge this PR as is, and then have a follow-up PR to address these bullets.

**Review by nilsvu — DISMISSED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**